### PR TITLE
Pure BatchNormalization

### PR DESCRIPTION
**Pull request by orausch**

### CODE REVIEW

File: daceml/onnx/op_implementations/img_op_implementations.py [hunk 1]
```diff
@@ -232,129 +241,172 @@ def forward(node: ONNXOp, state: SDFGState,
...
+        N = _prod(reduce_axes)
+        broadcast_shape = [num_channels, 1, 1]
+        dtype = in_desc_with_name(node, state, sdfg, "X").dtype
+        eps = node.epsilon
+        momentum = node.momentum
+        inv_momentum = 1 - node.momentum
+
+        def prog(X, scale, B, in_mean, in_var, Y, out_mean, out_var,
+                 saved_mean, saved_var):
+            saved_mean[:] = np.add.reduce(X, axis=[0, 2, 3]) / N
```
Comment (tbennun) [inline daceml/onnx/op_implementations/img_op_implementations.py:364]: these axes here should be computed outside (I think you can just use `reduce_axes`)

File: daceml/onnx/op_implementations/img_op_implementations.py [hunk 2]
```diff
@@ -232,129 +241,172 @@ def forward(node: ONNXOp, state: SDFGState,
...
+            saved_mean_broadcastable = dace.define_local(
+                broadcast_shape, dtype)
+            # this copy will get removed after parsing -- using reshape here would be nicer
+            # but it messes with statefusion
+            saved_mean_broadcastable[:] = saved_mean
+
+            X_minus_mean = (X - saved_mean_broadcastable)
+
+            saved_var[:] = np.add.reduce(X_minus_mean * X_minus_mean,
+                                         axis=(0, 2, 3)) / N
```
Comment (tbennun) [inline daceml/onnx/op_implementations/img_op_implementations.py:375]: same

File: daceml/onnx/schema.py [hunk 3]
```diff
@@ -93,7 +93,9 @@ def from_json(cls, json, parent=None):
             return cls(**constructor_args)
 
         def to_json(self):
-            return dace.serialize.all_properties_to_json(self)
+            serialized = dace.serialize.all_properties_to_json(self)
+            serialized["type"] = cls.__name__
```
Comment (tbennun) [inline daceml/onnx/schema.py:97]: still necessary?

File: daceml/util/utils.py [hunk 4]
```diff
@@ -122,13 +124,14 @@ def expand_onnx_nodes(sdfg: dace.SDFG):
...
-                impl_name = node.expand(sdfg, state)
-                print(
-                    "Automatically expanded library node \"{}\" with implementation \"{}\"."
-                    .format(str(node), impl_name))
-                # We made a copy of the original list of nodes, so we keep
-                # iterating even though this list has now changed
-                expanded_something = True
+                if predicate is None or predicate(node):
+                    impl_name = node.expand(sdfg, state)
+                    print(
```
Comment (tbennun) [inline daceml/util/utils.py:129]: if config.get_bool('debugprint')

**Review by tbennun — CHANGES_REQUESTED** [review]

File: daceml/onnx/schema.py (hunk 3 above)
Comment (orausch) [reply daceml/onnx/schema.py:97]: We don't use it, but it does improve the situation

**Review by tbennun — APPROVED** [review]

**Pull request merged**